bpf/lib/nodeport.h

```diff
@@ -188,13 +188,19 @@ static __always_inline int nodeport_nat_ipv6_fwd(struct __ctx_buff *ctx,
 		.max_port = NODEPORT_PORT_MAX_NAT,
 	};
 	int ret;
+	__u32 monitor = 0;
 
 	ipv6_addr_copy(&target.addr, addr);
 
 	ret = snat_v6_needed(ctx, addr) ?
 	      snat_v6_process(ctx, NAT_DIR_EGRESS, &target) : CTX_ACT_OK;
+
 	if (ret == NAT_PUNT_TO_STACK)
-		ret = CTX_ACT_OK;
+		return CTX_ACT_OK;
```

Comment:
I hope this is okay to do. From my understanding, `NAT_PUNT_TO_STACK` is returned from NAT codepaths to indicate that a packet should be processed by another BPF program, but I still have no clue as to how that works or why it is necessary. Where can I learn more about this?

Reply:
`git blame` is your friend. Running git blame on lines of code with `NAT_PUNT_TO_STACK` points towards commits 428f43e177de53fe7da0fd885da2eb577fdb071d and efd1ad80bd4ed20388a968185dfe13f920fd1196 with more context around the introduction of the `NAT_PUNT_TO_STACK` return code.

Though I don't understand why the early return in this case, don't we want to send_trace_notify() even in the case that snat_v6_process() returns `NAT_PUNT_TO_STACK`?

Reply:
> `git blame` is your friend

Thanks, hadn't noticed those other occurrences. `git blame` on the occurrences in `nodeport.h` didn't yield anything particularly useful. To which extent do we consider git history documentation? As a codebase goes through refactors, or when those particular lines get touched in any way, the paper trail starts to fade.

The commit messages do provide the proper context and the reason behind the change, but don't answer the question where packet processing eventually continues after they've been passed to the stack.

> don't we want to call `send_trace_notify()` even in the case that `snat_v6_process()` returns `NAT_PUNT_TO_STACK`?

That's exactly what I can't figure out. Does punting to the stack mean a packet is put on the wire? If that's not a 100% sure, we can't log a `TRACE_TO_NETWORK` here, since further processing might mean the packet gets steered elsewhere.

Reply:
I've done some testing with the early return, and it looks like `NAT_PUNT_TO_STACK` get set in the majority of cases, causing the trace to not being emitted. That change has been reverted.